File: frontend/api-processor/src/types/internal/operation.ts
/**
 * Copyright 2024-2025 NetCracker Technology Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DeprecateItem, OperationsApiType } from '../external'
import { ApiAudience } from '../package'
import { OpenAPIV3 } from 'openapi-types'
import { GraphApiSchema } from '@netcracker/qubership-apihub-graphapi'
import { Realm } from '@netcracker/qubership-apihub-ddlapi'
import { ApihubApiCompatibilityKind } from '../../consts'
import { v3 as AsyncAPIV3 } from '@asyncapi/parser/esm/spec-types'

export interface OperationSearch {
  useOperationDataAsSearchText: boolean
  searchTextFilePath?: string
}

export interface ApiOperation<T = any, M = any> {
  operationId: string
  documentId: string
  apiType: OperationsApiType
  apiKind: ApihubApiCompatibilityKind
  deprecated: boolean
  tags: string[]
  metadata: M
  data?: T

  // new properties
  deprecatedItems?: DeprecateItem[]     // deprecated items
  deprecatedInfo?: string
  deprecatedInPreviousVersions?: string[]
  models?: Record<string, string>       // schema models { name: hash }

  // other params (not used in builder logic)
  // [key: string]: unknown

  title: string
  search: OperationSearch
  searchText?: string
  // refPackage?: PackageRef
  // changes?: OperationChanges[]
  // changeSummary?: ChangeSummary
  hasExample?: boolean
  apiAudience?: ApiAudience
  versionInternalDocumentId: string
}

// `Realm` is included so the DDL version-internal / comparison documents serialize through the same
// `serializeDocument` / `denormalize` path as REST (AD3; plan Tasks 3 & 8).
export type ApiDocument = OpenAPIV3.Document | GraphApiSchema | AsyncAPIV3.AsyncAPIObject | Realm

File: backend/agents-backend/view/operation.go
package view

type RestOperations struct {
	Operations []RestOperationView          `json:"operations"`
	Packages   map[string]PackageVersionRef `json:"packages,omitempty"`
}

type RestOperationView struct {
	OperationListView
	RestOperationMetadata
}

type OperationListView struct {
	CommonOperationView
	PackageRef string                 `json:"packageRef,omitempty"`
	Data       map[string]interface{} `json:"data,omitempty"`
}

type CommonOperationView struct {
	OperationId string `json:"operationId"`
	Title       string `json:"title"`
	DataHash    string `json:"dataHash"`
	Deprecated  bool   `json:"deprecated,omitempty"`
	ApiKind     string `json:"apiKind"`
	ApiType     string `json:"apiType"`
	ApiAudience string `json:"apiAudience"`
}

type RestOperationMetadata struct {
	Path   string   `json:"path"`
	Method string   `json:"method"`
	Tags   []string `json:"tags,omitempty"`
}
